File: fe/lexer.cpp
#include "lexer.hpp"

static int String_StripFront(std::string &str) {
    size_t start = 0;
    while (start < str.size() && std::isspace(static_cast<unsigned char>(str[start]))) {
        ++start;
    }
	str = str.substr(start);
	return start;
}

static bool String_StartsWith(const std::string &str, const std::string &prefix) {
	if (str.size() < prefix.size()) return false;
	for (size_t i = 0; i < prefix.size(); ++i) {
		if (str[i] != prefix[i]) return false;
	}
	return true;
}

static std::vector<std::pair<std::string, Token::Type>> operatorTable = {
	{"==" , Token::Type::Eq},
	{"="  , Token::Type::Equal},
	{"+"  , Token::Type::Plus},
	{"-"  , Token::Type::Minus},
	{"*"  , Token::Type::Mul},
	{"/"  , Token::Type::Div},
	{"!=" , Token::Type::Neq},
	{">=" , Token::Type::Ge},
	{"<=" , Token::Type::Le},
	{">"  , Token::Type::Gt},
	{"<"  , Token::Type::Lt},
	{","  , Token::Type::Comma},
	{"("  , Token::Type::LParen},
	{")"  , Token::Type::RParen},
};

static std::vector<std::pair<std::string, Token::Type>> keywordTable = {
	{"function" , Token::Type::Function},
	{"arg"		, Token::Type::Arg},
	{"goto"		, Token::Type::Goto},
	{"label"	, Token::Type::Label},
	{"if"		, Token::Type::If},
	{"param"	, Token::Type::Param},
	{"call"		, Token::Type::Call},
	{"ret"		, Token::Type::Ret},
};

std::ostream &operator<<(std::ostream &os, const Loc &loc) {
	os << loc.filename << ":" << loc.row << ":" << loc.col;
	return os;
}

std::string Token::TypeToString() const {
	switch (this->type) {
	case Token::Type::Function: return "Function";
	case Token::Type::Iden: return "Iden";
	case Token::Type::Num: return "Num";
	case Token::Type::Comma: return "Comma";
	case Token::Type::Arg: return "Arg";
	case Token::Type::Equal: return "Equal";
	case Token::Type::Plus: return "Plus";
	case Token::Type::Minus: return "Minus";
	case Token::Type::Mul: return "Mul";
	case Token::Type::Div: return "Div";
	case Token::Type::Goto: return "Goto";
	case Token::Type::Label: return "Label";
	case Token::Type::If: return "If";
	case Token::Type::LParen: return "LParen";
	case Token::Type::RParen: return "RParen";
	case Token::Type::Eq: return "Eq";
	case Token::Type::Neq: return "Neq";
	case Token::Type::Gt: return "Gt";
	case Token::Type::Lt: return "Lt";
	case Token::Type::Ge: return "Ge";
	case Token::Type::Le: return "Le";
	case Token::Type::Param: return "Param";
	case Token::Type::Call: return "Call";
	case Token::Type::Ret: return "Ret";
	case Token::Type::Eol: return "Eol";
	case Token::Type::Eof: return "Eof";
	default:
        std::cerr << Func_Loc << std::endl << this->loc << ": error: "
				  << "encountered invalid token type" << std::endl;
		exit(1);
	}
}

void Token::ErrorTypeMismatch(const std::vector<Token::Type> types) const {
    switch (types.size()) {
	case 0:
		std::cerr << this->loc << ": error: expected no token; but got token of type '"
				  << this->TypeToString() << "'" << std::endl;
		break;
	case 1:
		std::cerr << this->loc << ": error: expected token of type '"
				  << (Token){.type=types[0]}.TypeToString() << "'; but got '"
				  << this->TypeToString() << "'" << std::endl;
		break;
	default:
		std::cerr << this->loc << ": error: expected token of types '";
		for (size_t i = 0; i < types.size(); ++i) {
			std::cerr << (Token){.type=types[i]}.TypeToString();
			if (i < types.size()-1) std::cerr << " | "; 
		}
		std::cerr << "'; but got '" << this->TypeToString() << "'" << std::endl;
	}
	exit(1);
} 

void Token::AssertTokenType(const Type type) const {
	if (this->type == type) return;
	this->ErrorTypeMismatch({type});
}

bool Token::IsArith() const {
	switch (this->type) {
	case Token::Type::Plus: return true;
	case Token::Type::Minus: return true;
	case Token::Type::Mul: return true;
	case Token::Type::Div: return true;
	default: return false;
	}
}

bool Token::IsUnary() const {
	switch (this->type) {
	case Token::Type::Minus: return true;
	default: return false;
	} 
}

bool Token::IsRelop() const {
	switch (this->type) {
	case Token::Type::Eq: return true;
	case Token::Type::Neq: return true;
	case Token::Type::Gt: return true;
	case Token::Type::Lt: return true;
	case Token::Type::Le: return true;
	case Token::Type::Ge: return true;
	default: return false;
	} 
}

Lexer::Lexer(std::string &filename) {
	std::ifstream file(filename);
    if (file.fail()) {
        std::cerr << Func_Loc << std::endl << "Error: Filename '"
				  << filename << "' does not exist" << std::endl;
        exit(1);
    }
    std::string line;
    while (getline(file, line)) this->content.push_back(line);
    file.close();

    if (this->content.empty()) return;
	size_t row, col;
    for (size_t i = 0; i < this->content.size(); ++i) {
        row = i+1, col = 1;
		std::string line(this->content[i]);
		
        while (!line.empty()) {
            col += String_StripFront(line);
            if (line.empty()) continue;
			
            bool flag = false;
            for (std::pair<std::string, Token::Type> op: operatorTable) {
                if (String_StartsWith(line, op.first)) {
                    this->tokens.push_back((Token){
							.type = op.second,
							.content = op.first,
							.loc  = (Loc){.row = row, .col = col, .filename=filename}
						});
                    line = line.substr(op.first.size());
                    col += op.first.size();
                    flag = true;
                    break;
                }
            }
            if (flag) continue;

			if (std::isalpha(static_cast<unsigned char>(line[0]))) {
				size_t j = 1;
				while (j < line.size() && std::isalnum(static_cast<unsigned char>(line[j]))) {
					++j;
				}
				std::string iden = line.substr(0, j);
				Token::Type type = Token::Type::Iden;
				for (std::pair<std::string, Token::Type> op: keywordTable) {
					if (iden.compare(op.first) == 0) {
						type = op.second;
						break;
					}
				}
				this->tokens.push_back((Token){
						.type = type,
						.content = iden,
						.loc  = (Loc){.row = row, .col = col, .filename=filename}
					});
				line = line.substr(iden.size());
				col += iden.size();
				continue;
			}
			
			if (std::isdigit(static_cast<unsigned char>(line[0]))) {
				size_t j = 1;
				while (j < line.size() && std::isdigit(static_cast<unsigned char>(line[j]))) {
					++j;
				}
				std::string iden = line.substr(0, j);
				this->tokens.push_back((Token){
						.type = Token::Type::Num,
						.content = iden,
						.loc  = (Loc){.row = row, .col = col, .filename=filename}
					});
				line = line.substr(iden.size());
				col += iden.size();
				continue;
			}

			Loc loc = (Loc){.row=row, .col=col, .filename=filename};
            std::cerr << Func_Loc << std::endl << loc << ": error: unknown token '"
					  << line[0] << "'" << std::endl;
            exit(1);
        }

		this->tokens.push_back((Token){
				.type = Token::Type::Eol,
				.content = "Eol",
				.loc  = (Loc){.row = row, .col = col, .filename=filename}
			});
    }
	
	this->tokens.push_back((Token){
			.type = Token::Type::Eof,
			.content = "Eof",
			.loc  = (Loc){.row = row, .col = col, .filename=filename}
		});
}

std::ostream &operator<<(std::ostream &os, const Lexer &lexer) {
	for (Token token: lexer.tokens) {
		os << token.TypeToString() << " \t " << token.content << std::endl;
	}
	return os;
}

File: src/fe/lexer.rs
use std::io::{prelude::*, BufReader};
use std::collections::HashMap;
use crate::fe::{token::{Token, TokenKind}, loc::Loc};

pub struct Lexer {
	pub tokens: Vec<Token>
}

impl Lexer {
	pub fn new(filepath: String) -> Self {
		let mut tokens: Vec<Token> = Vec::new();
		let mut content: Vec<String> = BufReader
			::new(std::fs::File::open(&filepath).expect("could not open file"))
			.lines()
			.map(|line| line.expect("could not parse file"))
			.collect();
		
		if content.is_empty() {
			return Self{tokens};
		}

		let mut row: usize = 0;
		let mut col: usize = 0;

		let operator_table: Vec<(String, TokenKind)> = vec![
			(String::from("=="), TokenKind::Eq),
			(String::from("="), TokenKind::Equal),
			(String::from("+"), TokenKind::Plus),
			(String::from("-"), TokenKind::Minus),
			(String::from("*"), TokenKind::Mul),
			(String::from("//"), TokenKind::Comment),
			(String::from("/"), TokenKind::Div),
			(String::from("!="), TokenKind::Neq),
			(String::from(">="), TokenKind::Ge),
			(String::from("<="), TokenKind::Le),
			(String::from(">"), TokenKind::Gt),
			(String::from("<"), TokenKind::Lt),
			(String::from(","), TokenKind::Comma),
			(String::from("("), TokenKind::LParen),
			(String::from(")"), TokenKind::RParen),
		];

		let keyword_table: HashMap<String, TokenKind> = [
			(String::from("function"), TokenKind::Function),
		    (String::from("arg"), TokenKind::Arg),
		    (String::from("goto"), TokenKind::Goto),
		    (String::from("label"), TokenKind::Label),
		    (String::from("if"), TokenKind::If),
		    (String::from("param"), TokenKind::Param),
		    (String::from("call"), TokenKind::Call),
		    (String::from("ret"), TokenKind::Ret)
		].iter().cloned().collect();

		for i in 0..content.len() {
			row = i+1;
			col = 1;

			while !content[i].is_empty() {

				let mut l: usize = 0;
				for c in content[i].chars() {
					if c.is_whitespace() {
						l += 1;
					} else {
						break;
					}
				}
				content[i] = content[i][l..].to_string();
				if content[i].is_empty() {
					break;
				}
				col += l;
				let loc: Loc = Loc::new(row, col, filepath.clone());

				let mut flag: bool = false;
				for (operator, token_kind) in &operator_table {
					if content[i].starts_with(operator) {
						tokens.push(Token::new(token_kind.clone(), loc.clone()));
						content[i] = content[i][operator.len()..].to_string();
						col += operator.len();
						flag = true;
						break;
					}
				}
				if flag {
					continue;
				}

				if content[i].chars().nth(0).unwrap().is_alphabetic() {
					let mut keyword_sz: usize = 1;
					while keyword_sz < content[i].len() &&
						content[i].chars().nth(keyword_sz).unwrap().is_alphanumeric() {
						keyword_sz += 1;
					}
					let keyword: String = content[i][0..keyword_sz].to_string();
					match keyword_table.get(&keyword) {
						Some(token_kind) => tokens.push(Token::new(token_kind.clone(), loc.clone())),
						None => tokens.push(Token::new(TokenKind::Iden(keyword), loc.clone())),
					}
					content[i] = content[i][keyword_sz..].to_string();
					col += keyword_sz;
					continue;
				}

				if content[i].chars().nth(0).unwrap().is_numeric() {
					let mut num_sz: usize = 0;
					let mut num: i32 = 0;
					while num_sz < content[i].len() && content[i].chars().nth(num_sz).unwrap().is_numeric() {
						num = num*10 + content[i].chars().nth(num_sz).unwrap() as i32 - '0' as i32;
						num_sz += 1;
					}
					tokens.push(Token::new(TokenKind::Num(num), loc.clone()));
					content[i] = content[i][num_sz..].to_string();
					col += num_sz;
					continue;
				}

				loc.error(format!("unexpected token '{}'", content[i].chars().nth(0).unwrap()));
			}
			tokens.push(Token::new(TokenKind::Eol, Loc::new(row, col, filepath.clone())));
		}
		tokens.push(Token::new(TokenKind::Eof, Loc::new(row, col, filepath.clone())));

		return Self{tokens};
	}
}
